test/tlsconnection.js

```diff
@@ -1119,11 +1461,48 @@ describe('the handshake between ClientConnection and ServerConnection', () => {
     assert.equal(SERVER_SENT.length, 3);
     await client.recv(SERVER_SENT[0]);
     await client.close();
-    assert.equal(CLIENT_SENT.length, 3);
-    await server.recv(CLIENT_SENT[1]);
-    await server.recv(CLIENT_SENT[2]);
+    assert.equal(CLIENT_SENT.length, 2);
+    assert.throwsAsync(async () => {
+      await server.recv(CLIENT_SENT[1]);
+    }, TLSCloseNotify);
     await server.close();
-    assert.equal(SERVER_SENT.length, 5);
+    assert.equal(SERVER_SENT.length, 4);
+    assert.throwsAsync(async () => {
+      await client.recv(SERVER_SENT[3]);
+    }, TLSCloseNotify);
+  });
+
+  it('rejects the `connected` promise on failure', async () => {
+    const CLIENT_SENT = [], SERVER_SENT = [];
+    const server = await ServerConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => SERVER_SENT.push(data));
+    const client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+    await assert.throwsAsync(async () => {
+      await server.recv(utf8ToBytes('junk data'));
+    }, TLSError, 'DECODE_ERROR');
+    // The server's `connected` promise rejects with that error.
+    await assert.throwsAsync(async () => {
+      await server.connected;
+    }, TLSError, 'DECODE_ERROR');
+    // The server sends the error to the client, which also rejects with it.
+    await assert.throwsAsync(async () => {
+      await client.recv(SERVER_SENT[0]);
+    }, TLSError, 'DECODE_ERROR');
+    await assert.throwsAsync(async () => {
+      await client.connected;
+    }, TLSError, 'DECODE_ERROR');
+  });
+
+  it('fulfills the `connected` promise on success', async () => {
+    const CLIENT_SENT = [], SERVER_SENT = [];
+    const server = await ServerConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => SERVER_SENT.push(data));
+    const client = await ClientConnection.create(TEST_VECTORS.PSK, TEST_VECTORS.PSK_ID, data => CLIENT_SENT.push(data));
+    await server.recv(CLIENT_SENT[0]);
+    await client.recv(SERVER_SENT[0]);
+    await client.recv(SERVER_SENT[1]);
+    await client.recv(SERVER_SENT[2]);
+    await server.recv(CLIENT_SENT[1]);
+    await client.connected;
```

Comment:
You can test that a promise is not fulfilled with Promise.any() if you are interested in testing that this isn't fulfilled earlier.